File: 98. Validate Binary Search Tree/isValidBST.js
// Definition for a binary tree node.
class TreeNode {
  constructor(val = 0, left = null, right = null) {
    this.val = val;
    this.left = left;
    this.right = right;
  }
}
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
const isValidBST = (root) => {
  const validate = (node, low = -Infinity, high = Infinity) => {
    if (!node) {
      return true;
    }

    if (node.val <= low || node.val >= high) {
      return false;
    }

    return validate(node.left, low, node.val) && validate(node.right, node.val, high);
  }

  return validate(root);
};


const root = new TreeNode(5);
root.left = new TreeNode(1);
root.right = new TreeNode(4);

root.right.left = new TreeNode(3);
root.right.right = new TreeNode(6);

console.log('isValidBST: ', isValidBST(root));

File: 98. Validate Binary Search Tree/isValidBST.py
# Definition for a binary tree node.
import math
from typing import Optional

class TreeNode:
  def __init__(self, val=0, left=None, right=None):
    self.val = val
    self.left = left
    self.right = right

class Solution:
  def isValidBST(self, root: Optional[TreeNode]) -> bool:
    def validate(node, low = float('-inf'), high = float('inf'),):
      if not node:
        return True
      
      if node.val <= low or node.val >= high:
        return False
      
      return validate(node.left, low, node.val) and validate(node.right, node.val, high)

    return validate(root)
    

root = TreeNode(5)
root.left = TreeNode(1)
root.right = TreeNode(4)

root.right.left = TreeNode(3)
root.right.right = TreeNode(6)

solution = Solution()
print('isValidBST: ', solution.isValidBST(root))
